python/cudf/cudf/_lib/scalar.pyx

```diff
@@ -93,41 +120,73 @@ cdef class Scalar:
                 f"{type(value).__name__} to cudf scalar"
             )
 
+    def _get_device_value(self):
+        if pd.api.types.is_string_dtype(self.dtype):
+            result = _get_py_string_from_string(self.c_value)
+        elif pd.api.types.is_numeric_dtype(self.dtype):
+            result = _get_np_scalar_from_numeric(self.c_value)
+        elif pd.api.types.is_datetime64_dtype(self.dtype):
+            result = _get_np_scalar_from_timestamp64(self.c_value)
+        elif pd.api.types.is_timedelta64_dtype(self.dtype):
+            result = _get_np_scalar_from_timedelta64(self.c_value)
+        else:
+            raise ValueError(
+                "Could not convert cudf::scalar to a Python value"
+            )
+        return result
+
     @property
     def dtype(self):
         """
         The NumPy dtype corresponding to the data type of the underlying
         device scalar.
         """
-        cdef libcudf_types.data_type cdtype = self.c_value.get()[0].type()
+        cdef libcudf_types.data_type cdtype = self.get_raw_ptr()[0].type()
         return cudf_to_np_types[<underlying_type_t_type_id>(cdtype.id())]
 
     @property
     def value(self):
         """
         Returns a host copy of the underlying device scalar.
         """
-        if pd.api.types.is_string_dtype(self.dtype):
-            return _get_py_string_from_string(self.c_value)
-        elif pd.api.types.is_numeric_dtype(self.dtype):
-            return _get_np_scalar_from_numeric(self.c_value)
-        elif pd.api.types.is_datetime64_dtype(self.dtype):
-            return _get_np_scalar_from_timestamp64(self.c_value)
-        elif pd.api.types.is_timedelta64_dtype(self.dtype):
-            return _get_np_scalar_from_timedelta64(self.c_value)
+        if self._is_host_value_current():
+            return self._host_value
         else:
-            raise ValueError(
-                "Could not convert cudf::scalar to a Python value"
-            )
+            result = self._get_device_value()
+            self._host_value = result
+            return result
+
+    def _sync(self):
+        """
+        If the cache is not synched, copy either the device or host value
+        to the host or device respectively. If cache is valid, do nothing
+        """
+        if self._is_host_value_current() and self._is_device_value_current():
+            return
+        elif (
+            self._is_host_value_current() and not
+            self._is_device_value_current()
+        ):
+            self._set_device_value(self._host_value, self._host_dtype)
+        elif (
+            self._is_device_value_current() and not
+            self._is_host_value_current()
+        ):
+            self._host_value = self._get_device_value()
```

Comment:
Since there should be not way to end up with both host and device values being invalid, we might want to make that explicit. 
```suggestion
        if self._is_host_value_current() and self._is_device_value_current():
            return

        if self.is_host_value_current:
            self._set_device_value(self._host_value, self._host_dtype)
            return
            
        if self._is_device_value_current():
            self._host_value = self._get_device_value()
            return
        
        # we should never get here.
        raise Error("something about undefined behavior") 
```

Reply:
I think this is a good suggestion. My only fear is that we don't actually go through `sync` to serve the results from user facing APIs, so it won't protect them from creating an invalid state. For instance the `value` property is written as 

```
        if self._is_host_value_current():
            return self._host_value
        else:
            result = self._get_device_value()
            self._host_value = result
            return result
```

If we put a `_sync()` in there, then someone creating a host side scalar would inadvertently copy it to device without needing to just to call `value`.